File: src/main/kotlin/uk/ac/ic/matterialize/camera/OpenCVWebcam.kt
package uk.ac.ic.matterialize.camera

import org.opencv.core.Mat
import org.opencv.core.MatOfByte
import org.opencv.imgcodecs.Imgcodecs
import org.opencv.videoio.VideoCapture
import org.opencv.videoio.Videoio.CAP_V4L
import java.awt.image.BufferedImage
import java.io.ByteArrayInputStream
import javax.imageio.ImageIO

class OpenCVWebcam(private val device: Int, private val width: Int, private val height: Int) {
    companion object {
        init {
            nu.pattern.OpenCV.loadLocally()
        }

        fun convertToBufferedImage(mat: Mat): BufferedImage {
            val mob = MatOfByte()
            Imgcodecs.imencode(".png", mat, mob)
            return ImageIO.read(ByteArrayInputStream(mob.toArray()))
        }
    }

    private var capture: VideoCapture? = null

    fun start() {
        capture = VideoCapture()
        capture!!.open(device, CAP_V4L)
    }

    fun stop() {
        capture?.release()
        capture = null
    }

    fun grab(): Mat {
        val mat = Mat()
        // TODO add proper error
        capture!!.read(mat)
        return mat
    }

    fun fps(samples: Int): Double {
        println("Taking $samples samples to estimate FPS...")
        var sum = 0L

        (0 until samples).forEach {
            val start = System.currentTimeMillis()
            grab()
            val end = System.currentTimeMillis()

            sum += end - start
        }

        return 1000 / (sum / samples.toDouble())
    }
}

File: src/camera/opencv_webcam.cpp
#include "opencv_webcam.hpp"

#include <chrono>
#include <iostream>

#include <opencv2/core.hpp>
#include <opencv2/videoio.hpp>

void OpenCVWebcam::start() {
  capture.open(device, cv::CAP_V4L);
  capture.set(cv::CAP_PROP_FRAME_WIDTH, width);
  capture.set(cv::CAP_PROP_FRAME_HEIGHT, height);
  //  capture.set(cv::CAP_PROP_BUFFERSIZE, 3);
}

void OpenCVWebcam::stop() {
  capture.release();
}

cv::Mat OpenCVWebcam::grab() {
  cv::Mat frame{};

  capture >> frame;

  return frame;
}

long double OpenCVWebcam::fps(int samples) {
  std::cout << "Taking " << samples << " samples to estimate FPS...\n";

  long sum{};

  for (int i{}; i < samples; ++i) {
    auto start{std::chrono::system_clock::now()};

    grab();

    auto end{std::chrono::system_clock::now()};

    sum += std::chrono::duration_cast<std::chrono::milliseconds>(end - start)
               .count();
  }

  return 1000 / (sum / static_cast<long double>(samples));
}
